File: C++/isValidBst.cpp
#include<iostream>
#include<vector>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        std::vector<int>v=inorder(root);
        for(int i=1;i<v.size();i++){
            if(v[i-1]>v[i]){
                return false;
            }
        }
        return true;
    }
    std::vector<int>inorder(TreeNode* root){
        if(!root){return std::vector<int>();}
        std::vector<int>v;
        std::vector<int>left=inorder(root->left);
        v.insert(v.end(),left.begin(),left.end());
        v.push_back(root->val);
        std::vector<int>right=inorder(root->right);
        v.insert(v.end(),right.begin(),right.end());
        return v;
    }
};

void deleteTree(TreeNode* root){
    if(root){
        deleteTree(root->left);
        deleteTree(root->right);
        delete root;
    }
}

int main(){
    TreeNode* root=new TreeNode(1);
    root->left=new TreeNode(2);
    root->right=new TreeNode(3);
    Solution solution;
    bool isBST=solution.isValidBST(root);

    if(isBST){
        std::cout<<"The given binary tree is a valid binary search tree."<<std::endl;
    }
    else{
        std::cout<<"The given binary tree is not a valid binary search tree."<<std::endl;
    }

    deleteTree(root);

    return 0;
}

File: python/isValidBST.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isValidBST(self , root: TreeNode) -> bool:
        v = self.inorderTraversal(root)
        for i in range(1, len(v)):
            if v[i - 1] >= v[i]:
                return False
        return True

    def inorderTraversal(self, root: TreeNode) -> list[int]:
        v = []

        def inorder(node):
            if node is None:
                return
            inorder(node.left)
            v.append(node.val)
            inorder(node.right)

        inorder(root)
        return v

root1=TreeNode(1)
root1.left=TreeNode(2)
root1.right=TreeNode(3)

root2=TreeNode(2)
root2.left=TreeNode(1)
root2.right=TreeNode(3)

solution=Solution()
list1=solution.isValidBST(root1)
list2=solution.isValidBST(root2)

print(list1)
print(list2)